### ISSUE TITLE
JDBC OutOfMemoryError when reading TEXT[] arrays a specific way

### ISSUE THREAD

**Issue by alex-sherwin** [cockroachdb/cockroach#20372]
**BUG REPORT**

1. Please supply the header (i.e. the first few lines) of your most recent
   log file **for each node in your cluster**. On most unix-based systems
   running with defaults, this boils down to the output of

Simple vanilla instance of `1.1.3`

```
CockroachDB node starting at 2017-12-01 04:41:56.66538826 +0000 UTC (took 1.2s)
build:      CCL v1.1.3 @ 2017/11/27 13:59:10 (go1.8.3)
admin:      http://test1:8080
sql:        postgresql://root@test1:26257?application_name=cockroach&sslmode=disable
logs:       /cockroach/cockroach-data/logs
store[0]:   path=/cockroach/cockroach-data
status:     initialized new cluster
clusterID:  3c5af0cd-d65f-4da5-b9c7-59d8f95c927c
nodeID:     1
```

2. Please describe the issue you observed:

- What did you do?

Read array column of type `TEXT[]` data in a specific pattern via Java/JDBC

- What did you expect to see?

Consistent working behavior (i.e. consistently reading the data from the selected data repeatedly)

- What did you see instead?

On exactly the 6th iteration of specific reading pattern, the Java application will `OutOfMemoryError` due to the Postgres JDBC `Field` object changing to format `binary`, which then causes the array size to be mis-read, which then tries to initialize an `int[]` array using the mis-read array size (which was mis-read as a huge value), causing an `OutOfMemoryError` trying to allocate a huge `int[]` array

I have no idea if this is a CockroachDB bug or a Postgres JDBC driver bug

I've created a complete working POC repo with instructions and sample application which exhibits this behavior, it can be setup and run very easily, please see https://github.com/alex-sherwin/crdb-oom

**Comment by alex-sherwin** [comment]
I should note, if you explicitly set JDBC URL parameter `binaryTransfer=false` then this won't occur (i.e. the JDBC driver will never switch to `binary` mode and then this problem doesn't occur)

While it's a workaround, I don't think it's a proper solution, this definitely seems like a bug somewhere in either CRDB or the Postgres JDBC driver

**Comment by jseldess** [comment]
Thanks for reporting this, @alex-sherwin. I'll find a dev on our end to help out.

**Comment by justinj** [comment]
@alex-sherwin thank you SO much for the detailed repro. I'm certain this was causing #19669 as well but I couldn't figure out the cause. Your repo was extremely helpful in figuring out the cause, and this is fixed by #20461.

**Comment by alex-sherwin** [comment]
Great to hear @justinj, glad I could help
